src/main/java/mods/railcraft/common/blocks/tracks/TrackSwitch.java

```diff
@@ -44,7 +46,7 @@ public IIcon getIcon() {
     @Override
     public int getBasicRailMetadata(EntityMinecart cart) {
         int meta = tileEntity.getBlockMetadata();
-        if (cart != null && isSwitched()) {
+        if (cart != null && (isSwitched() || springingCarts.contains(cart.getPersistentID()))) {
```

Comment:
Wouldn't this "derail" a cart from the "locked" direction if the track was switched? Probably need to add a check against the locked set when isSwitched() is true.